unumpy/linalg/_multimethods.py

```diff
@@ -0,0 +1,203 @@
+import functools
+import operator
+from uarray import create_multimethod, mark_as, all_of_type, Dispatchable
+import builtins
+
+create_numpy = functools.partial(create_multimethod, domain="numpy.linalg")
+
+from .._multimethods import (
+    ndarray,
+    _identity_argreplacer,
+    _self_argreplacer,
+    _dtype_argreplacer,
+    mark_dtype,
+    _first_argreplacer,
+    _first2argreplacer,
+)
+
+
+@create_numpy(_first_argreplacer)
+@all_of_type(ndarray)
+def multi_dot(arrays):
+    return arrays
+
+
+def matrix_power_default(a, n):
+    if not ndim(a) == 2:
+        raise ValueError("Array must be two-dimensional")
+
+    if not isinstance(n, int):
+        raise ValueError("Exponent must be an integer.")
+
+    if a.shape[0] == a.shape[1]:
+        if n == 0:
+            return eye(a.shape[0], a.shape[1])
+        elif n < 0:
+            a = inv(a)
+
+        eigenvalues, eigenvectors = eig(a)
+
+        diagonal = diag(eigenvalues)
+
+        return multi_dot([eigenvectors, power(diagonal, abs(n)), inv(eigenvectors)])
+    else:
+        raise ValueError("The array must be square.")
+
+
+@create_numpy(_self_argreplacer, default=matrix_power_default)
+@all_of_type(ndarray)
+def matrix_power(a, n):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def cholesky(a):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def qr(a, mode="reduced"):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def svd(a, full_matrices=True, compute_uv=True, hermitian=False):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def eig(a):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def eigh(a, UPLO="L"):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def eigvals(a):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def eigvalsh(a, UPLO="L"):
+    return (a,)
+
+
+def norm_default(x, ord=None, axis=None, keepdims=False):
```

Comment:
The axis argument is ignored

Reply:
Yes, some norms for matrices (2-norm and smallest singular value) are also ignored, I will try to add them as well if they are simple enough.

I'm still confused about how to use the axis argument but I will take a closer look. Can I assume only two-dimensional arrays are given? Because these might be easier for me.

Reply:
From the docs:

>If axis is an integer, it specifies the axis of x along which to compute the vector norms. If axis is a 2-tuple, it specifies the axes that hold 2-D matrices, and the matrix norms of these matrices are computed.

You've implemented the `axis=None` case. But if `isinstance(axis, int)` then you should do a vector norm over that axis,regardless of `ndim(x)`. That should be as simple as calling the underlyinh reduction operation (`sum`, `min` or `max`) with the `axis` and `keepdims` arguments.

You can also have `axis` be a pair `(i, j)` which gives the dimensions for a matrix norm. That would mean where you've used `axis=0` here, you should instead use `axis=axis[0]` and similarly `axis=1` becomes `axis=axis[1]`.

Reply:
This was very useful, thank you!